### PR TITLE
New Command: Get-CrmSPDocLoc

### PR DESCRIPTION
**Pull request by RichieRoxx**
Hi Arjan, as announced via linked in: I've added a new cmdlets Get-CRMSPDocLoc to easily retrieve the Urls of a SharePoint Document Location. I work a lot with CRM and SharePoint-Integrations and regularly make use of this. Happy to adapt anything so you can pull this into your repo. Looking forward to hearing from you.

### CODE REVIEW

**Comment by RichieRoxx** [comment]
I noticed I need to switch codesigning back on. can certainly do that, but would like to collect the whole of your feedback first.

File: AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+using System.Linq;
+using System.Management.Automation;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace AMSoftware.Crm.PowerShell.Commands.Administration
+{
+    [Cmdlet(VerbsCommon.Get, "CrmSPDocLoc", HelpUri = HelpUrlConstants.GetRoleHelpUrl, SupportsPaging = true, DefaultParameterSetName = "")]
+    [OutputType(typeof(Entity))]
+    public class GetSPDocLoc :CrmCmdlet
```
Comment (AMSoftwareNL) [inline AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs]: Use fullname for Classname and all Cmdlet classes end in 'Command'
Inherit from CrmOrganizationCmdlet

```suggestion
    public class GetSPDocucmentLocationCommand :CrmOrganizationCmdlet
```

File: AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Management.Automation;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace AMSoftware.Crm.PowerShell.Commands.Administration
+{
+    [Cmdlet(VerbsCommon.Get, "CrmSPDocLoc", HelpUri = HelpUrlConstants.GetRoleHelpUrl, SupportsPaging = true, DefaultParameterSetName = "")]
```
Comment (AMSoftwareNL) [inline AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs]: I prefer fullnames for my Cmdlets. Autocomplete in PowerShell will help with typing, and this way the comdlets are clear.
Add a HelpUrl in the HelpUrlConstants, or leave out if no help provided.
```suggestion
    [Cmdlet(VerbsCommon.Get, "CrmSPDocumentLocation", HelpUri = HelpUrlConstants.GetSPDocumentLocationHelpUrl, SupportsPaging = true, DefaultParameterSetName = "")]
```

File: AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    [OutputType(typeof(Entity))]
+    public class GetSPDocLoc :CrmCmdlet
+    {
+        [Parameter(Mandatory = true)]
```
Comment (AMSoftwareNL) [inline AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs]: Include validation on the parameter if it is mandatory.
```suggestion
        [Parameter(Mandatory = true)]
        [ValidateNotNullOrEmpty]
```

File: AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        public Guid[] Id { get; set; }
+
+        [ArgumentCompleter(typeof(EntityArgumentCompleter))]
+        [Parameter]
+        public string Entity { get; set; }
```
Comment (AMSoftwareNL) [inline AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs]: If not using this then remove. 
I would prefer to keep the parameter and add an additional one for EntityId (Guid[]) so one or more specific Records can be provided. And use the Id parameter in a separate ParameterSet to retrieve a specific DocumentLocation record.

File: AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+                    }
+                }
+                
+            }
+            
+        }
+
+        private void RetrieveUrlById(Guid id)
+        {
+            var sourceUrlRequest = new RetrieveAbsoluteAndSiteCollectionUrlRequest
```
Comment (AMSoftwareNL) [inline AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs]: Prefer generic OrganizationRequest to limit the need for the Messages namespace. See SetOwnerCommand.cs for example.
```suggestion
            OrganizationRequest sourceUrlRequest = new OrganizationRequest("RetrieveAbsoluteAndSiteCollectionUrl")
```

File: AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+            {
+                Target = new EntityReference("sharepointdocumentlocation", id)
+            };
+            var sourceUrlResponse = (RetrieveAbsoluteAndSiteCollectionUrlResponse)CrmContext.OrganizationProxy.Execute(sourceUrlRequest);
```
Comment (AMSoftwareNL) [inline AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs]: Use generic OrganizationResponse to limit the need for the Messages namespace. See GetLanguageCommand
And use the Repository to execute the request
```suggestion
            OrganizationResponse sourceUrlResponse = _repository.Execute(sourceUrlRequest);
```

File: AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+                    RetrieveUrlById(id);
+                }
+
+            }
+            else
+            {
+                foreach (var id in Id)
+                {
+                    QueryByAttribute query = new QueryByAttribute("sharepointdocumentlocation");
+                    query.AddAttributeValue("regardingobjectid", id);
```
Comment (AMSoftwareNL) [inline AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs]: Include Entity and EntityId in query. There won't be duplicate ID's but regarding is a lookup so includes the Entity, and is an extra check for the user.

**Review by AMSoftwareNL — COMMENTED** [review]
Good addition to the library. Made some remarks and suggestions to match with the existing commands.
Would suggest to make two ParameterSets. One to get the document location by ID, and one to get by Record (Entity and EntityId as separate parameters).

**Comment by AMSoftwareNL** [comment]
You can disable code signing to compile, test and the pull request. I have the key only in my local repository, and will add on build of a new release. It should all work without the assembly being signed. It's just a quality thing.

File: AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs (hunk 2 above)
Comment (RichieRoxx) [reply AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs]: accepted

File: AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs (hunk 3 above)
Comment (RichieRoxx) [reply AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs]: accepted

File: AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs (hunk 4 above)
Comment (RichieRoxx) [reply AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs]: accepted

File: AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs (hunk 5 above)
Comment (RichieRoxx) [reply AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs]: In general I'd prefer strongly typed requests over generic requests as you suggest. Could you please elaborate on why you'd rather not want to use the messages namespace? Thank you!

File: AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs (hunk 7 above)
Comment (RichieRoxx) [reply AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs]: not exactly sure what you mean by this. Please see the latest changes and check if this is already obsolete.

**Review by AMSoftwareNL — APPROVED** [review]

**Pull request merged**